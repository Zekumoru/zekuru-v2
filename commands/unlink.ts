import {
  ChatInputCommandInteraction,
  PermissionFlagsBits,
  SlashCommandBuilder,
} from 'discord.js';
import { createCommand } from '../types/DiscordCommand';
import channelLinks from '../cache/channelLinks';
import { IChannelLink } from '../models/ChannelLink';

const UnlinkOptions = {
  SOURCE_CHANNEL: 'source-channel',
  TARGET_CHANNEL: 'target-channel',
};

const data = new SlashCommandBuilder()
  .setName('unlink')
  .setDescription('Unlinks two translation channels.')
  .addChannelOption((option) =>
    option
      .setName(UnlinkOptions.TARGET_CHANNEL)
      .setDescription('The target channel to unlink.')
      .setRequired(true)
  )
  .addChannelOption((option) =>
    option
      .setName(UnlinkOptions.SOURCE_CHANNEL)
      .setDescription(
        'The source channel to unlink. If not provided, takes the current channel as the source.'
      )
  )
  .setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild);

export const unlinkChannel = async (
  channelLink: IChannelLink,
  channelId: string
) => {
  if (channelLink.links.find((link) => link.id === channelId)) {
    channelLink.links = channelLink.links.filter(
      (trChannel) => trChannel.id !== channelId
    );

    if (channelLink.links.length === 0) {
      // remove channelLink since it doesn't have any links anymore
      // because it cannot be called "channelLink" with no links
      await channelLinks.delete(channelLink.id);
    } else {
      await channelLinks.update(channelLink);
    }

    return true;
  }
  return false;
};

const execute = async (interaction: ChatInputCommandInteraction) => {
  const sourceChannelId =
    interaction.options.getChannel(UnlinkOptions.SOURCE_CHANNEL)?.id ??
    interaction.channelId;

  const targetChannelId = interaction.options.getChannel(
    UnlinkOptions.TARGET_CHANNEL
  )?.id;
  if (!targetChannelId) {
    await interaction.reply({
      content: `Please specify the target channel.`,
    });
    return;
  }

  const [sourceChLink, targetChLink] = await Promise.all([
    channelLinks.get(sourceChannelId),
    channelLinks.get(targetChannelId),
  ]);

  let errorMessage = '';
  if (sourceChannelId === targetChannelId) {
    errorMessage = `You cannot unlink <#${sourceChannelId}> with itself!`;
  } else if (sourceChLink == null && targetChLink == null) {
    errorMessage = `Both <#${sourceChannelId}> and <#${targetChannelId}> are not linked with any channels!`;
  } else if (sourceChLink == null) {
    errorMessage = `<#${sourceChannelId}> is not linked with any channels!`;
  } else if (targetChLink == null) {
    errorMessage = `<#${targetChannelId}> is not linked with any channels!`;
  }

  if (errorMessage || sourceChLink == null || targetChLink == null) {
    await interaction.reply({
      content: errorMessage,
    });
    return;
  }

  // unlink bidirectional
  let unlinked = (
    await Promise.all([
      unlinkChannel(sourceChLink, targetChannelId),
      unlinkChannel(targetChLink, sourceChannelId),
    ])
  ).reduce((v1, v2) => v1 || v2);

  await interaction.reply({
    content: unlinked
      ? `<#${sourceChannelId}> and <#${targetChannelId}> are now unlinked!`
      : `<#${sourceChannelId}> and <#${targetChannelId}> are already unlinked!`,
  });
};

export default createCommand({
  data,
  execute,
});
