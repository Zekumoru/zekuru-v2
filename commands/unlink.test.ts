import unlink from './unlink';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';
import {
  sampleChannelLinks,
  sampleTranslateChannels,
} from '../cache/__mocks__/sample-data';

jest.mock('../cache');

describe('/unlink command', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it(`should notify user that the command is only for servers`, async () => {
    const interaction = createMockInteraction({ guildId: undefined });

    await unlink.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user that they didn't provide a target channel`, async () => {
    const interaction = createMockInteraction({ channelId: '100' });

    await unlink.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it.each([
    ['000', '000'],
    ['000', '001'],
    ['000', '100'],
    ['100', '000'],
    ['100', '100'],
  ])(
    `[%s, %s] - should notify user appropriately with regards to the state of the provided source and target channels`,
    async (sourceId, targetId) => {
      const interaction = createMockInteraction();
      interaction.options.getChannel.mockReturnValueOnce({
        id: sourceId,
      } as any);
      interaction.options.getChannel.mockReturnValueOnce({
        id: targetId,
      } as any);

      await unlink.execute(interaction);

      expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    }
  );

  it(`should notify user if the provided channels are already unlinked`, async () => {
    const interaction = createMockInteraction({ channelId: '100' });
    interaction.options.getChannel.mockReturnValueOnce(null);
    interaction.options.getChannel.mockReturnValueOnce({ id: '200' } as any);

    await unlink.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should unlink`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValueOnce({ id: '100' } as any);
    interaction.options.getChannel.mockReturnValueOnce({ id: '200' } as any);
    sampleChannelLinks[0].links.push(sampleTranslateChannels[1] as any);
    sampleChannelLinks[1].links.push(sampleTranslateChannels[0] as any);

    await unlink.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(sampleChannelLinks[0].links).not.toContain(sampleChannelLinks[1]);
    expect(sampleChannelLinks[1].links).not.toContain(sampleChannelLinks[0]);
  });
});
