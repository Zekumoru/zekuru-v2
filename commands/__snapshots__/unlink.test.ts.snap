// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/unlink command [000, 000] - should notify user appropriately with regards to the state of the provided source and target channels 1`] = `
{
  "content": "You cannot unlink <#000> with itself!",
}
`;

exports[`/unlink command [000, 001] - should notify user appropriately with regards to the state of the provided source and target channels 1`] = `
{
  "content": "Both <#000> and <#001> are not translate channels!",
}
`;

exports[`/unlink command [000, 100] - should notify user appropriately with regards to the state of the provided source and target channels 1`] = `
{
  "content": "<#000> is not a translate channel!",
}
`;

exports[`/unlink command [100, 000] - should notify user appropriately with regards to the state of the provided source and target channels 1`] = `
{
  "content": "<#000> is not a translate channel!",
}
`;

exports[`/unlink command [100, 100] - should notify user appropriately with regards to the state of the provided source and target channels 1`] = `
{
  "content": "You cannot unlink <#100> with itself!",
}
`;

exports[`/unlink command should notify user if the provided channels are already unlinked 1`] = `
{
  "content": "<#100> and <#200> are already unlinked!",
}
`;

exports[`/unlink command should notify user that the command is only for servers 1`] = `
{
  "content": "This command is only available on servers.",
}
`;

exports[`/unlink command should notify user that they didn't provide a target channel 1`] = `
{
  "content": "Please specify the target channel.",
}
`;

exports[`/unlink command should unlink 1`] = `
{
  "content": "<#100> and <#200> are now unlinked!",
}
`;
